### PR TITLE
Basic Finance Math

### LINKED ISSUE #1863: [Feature] Add basic finance math functions to `EconomicEntity`
# What is the motivation for your feature request? Please describe what you want to use it for.
In order to start performing financial/economic calculations, we need to support a standard set of financial function.

# Describe the feature you'd like.
The following primary functions are probably adequate:
* `PV(n,i,F,A)` - calculates the present value of a series of payments, using discount rate `i` per time period, for future payment `F` that occurs `n` time periods in the future plus regular payments `A` that occurs each time period for `n` periods
* `FV(n,i,P,A)` - future value for same parameters as above
* `PMT(n,i,P,F)` - computes the regular payment for `n` time periods that is equivalent to the combination of a single payment of `P` immediately and a single payment `F` after `n` time periods, using a discount rate of `i` per time periods
* `PV(i,[A])` - calculates the present value of a series of payments described by the vector `[A]`, occurring over a time period defined by the length of `[A]`, with a discount rate of `i` per time periods

## Describe alternatives you've considered.
There could be other ways to structure this, but these 4 functions can become the building blocks of all other needs.

# How can this issue be closed?
These four functions can be added to `EconomicEntity` to be available to all agents.

### PR DESCRIPTION
**Pull request by dean-krueger**
# Summary of Changes

This PR adds a set of basic financial math functions to the Economic Entity class, which will hopefully be useful in covering most finical math operations for all future functions, though they may need to be expanded in the future.

# Related CEPs and Issues

This PR is related to:

- Closes #1863 

# Associated Developers

@gonuke

# Design Notes

@gonuke laid out the set of functions to be added in Issue #1863, and this is hopefully a fairly straightforward implementation of those functions.

# Testing and Validation

Tests were added to agent_tests for each of these new functions where they are validated by hand calcs. All tests passed after rebuilding cyclus, and I manually verified that all four new tests were called in that set of unit tests to be sure I did it right.

# Checklist

 - [x] Read the [Contributing to Cyclus](https://fuelcycle.org/kernel/contributing_to_cyclus.html) guide.
 - [x]  Compile and run locally.
 - [x]  Add or update tests.
 - [x]  Document if needed.
 - [x]  Follow style guidelines.
 - [x]  Update the changelog.
 - [x]  Address all review comments.
Reviewers, please refer to the Cyclus [Guide for Reviewers](https://fuelcycle.org/kernel/pr_review.html).

### CODE REVIEW

**Comment by dean-krueger** [comment]
Screenshot 2025-06-14 at 11 36 38 AM

Here is a nicely formatted version of the functions I implemented, to hopefully make the review process a little easier. These are the forms I used to do my hand calculations in the tests.

File: tests/agent_tests/agent_tests.cc [hunk 1]
```diff
@@ -78,3 +78,27 @@ TEST_P(AgentTests, SetThenGetEconParam) {
   EXPECT_EQ(data, -1.0);
 }
 
+TEST_P(AgentTests, PV) {
+  double present_value = agent_->PV(10, 0.05, 10000.0, 100.0);
```
Comment (gonuke) [inline tests/agent_tests/agent_tests.cc:82]: Probably makes sense to test with only F, only A, with both, and with i = 0, i != 0, n<0, n=0, n>0

File: tests/agent_tests/agent_tests.cc [hunk 2]
```diff
@@ -78,3 +78,27 @@ TEST_P(AgentTests, SetThenGetEconParam) {
...
+  double pv_hand_calc = 6911.306028;
+  EXPECT_NEAR(present_value, pv_hand_calc, 1e-3);
+}
+
+TEST_P(AgentTests, FV) {
```
Comment (gonuke) [inline tests/agent_tests/agent_tests.cc:87]: ibid,regarding variety of testing

File: tests/agent_tests/agent_tests.cc [hunk 3]
```diff
@@ -78,3 +78,27 @@ TEST_P(AgentTests, SetThenGetEconParam) {
...
+  double future_value = agent_->FV(10, 0.05, 10000.0, 100.0);
+  double fv_hand_calc = 17546.73552;
+  EXPECT_NEAR(future_value, fv_hand_calc, 1e-3);
+}
+
+TEST_P(AgentTests, PMT) {
+  double payment = agent_->PMT(10, 0.05, 10000.0, 40000.0);
```
Comment (gonuke) [inline tests/agent_tests/agent_tests.cc:94]: This can be tested against PV and FV

PV(n, i, 0, PMT(n, i, 1, 0)) should equal 1

File: tests/agent_tests/agent_tests.cc [hunk 4]
```diff
@@ -78,3 +78,27 @@ TEST_P(AgentTests, SetThenGetEconParam) {
...
+  double pmt_hand_calc = 4475.2287;
+  EXPECT_NEAR(payment, pmt_hand_calc, 1e-3);
+}
+
+TEST_P(AgentTests, PV_def_by_A) {
+  std::vector<double> payments = {100.0, 200.0, 300.0, 10.0, 15.0, 1000.0};
```
Comment (gonuke) [inline tests/agent_tests/agent_tests.cc:100]: Can also be tested against prior functions

File: src/economic_entity.h [hunk 5]
```diff
@@ -32,6 +32,97 @@ class EconomicEntity {
...
+/// @param i Discount rate
+/// @param A Vector of payments which also defines how many time periods are
+/// summed over (by the length of A)
+/// @return Present value
+
+virtual double PV(double i, const std::vector<double>& A) const {
+    double pv = 0.0;
+
+    for (int t = 0; t < A.size(); ++t) {
+        pv += A[t] / std::pow((1 + i), t);
```
Comment (gonuke) [inline src/economic_entity.h:120]: There is almost certainly an off by one error of some kind here 

the contribution from the first payment `A[0]` should happen at the end of the first time period such that the PV factor is at `t=1`

You'd probably catch this if testing this function with a set of constant payments and comparing to PV

**Review by gonuke — CHANGES_REQUESTED** [review]
Thanks @dean-krueger.  I have a few comments on testing and one question about the implementation of the varying payments

**Comment by gonuke** [comment]
Regarding the formatted equations, the last one should go from t=[1,n] by convention

File: src/economic_entity.h (hunk 5 above)
Comment (dean-krueger) [reply src/economic_entity.h:120]: Ahh, that was a slight misunderstanding on my part, which is reflected in the way I was testing (my hand calc has the first exponent as a 0 not a 1). I'll fix this and then add those extra tests you were talking about to hopefully catch anything else like this.

**Comment by dean-krueger** [comment]
Not sure if the conda/Mac test is still failing. It did last time, but I've gone and added a bunch of those tests you asked for (good idea by the way, more tests is almost always better), and fixed the off-by-one error. Let me know if there are other things you see in your review, and I'll work on fixing them ASAP. If that one workflow test is still failing I can investigate why later.

File: tests/agent_tests/agent_tests.cc [hunk 6]
```diff
@@ -78,3 +79,93 @@ TEST_P(AgentTests, SetThenGetEconParam) {
   EXPECT_EQ(data, -1.0);
 }
 
+TEST_P(AgentTests, PV_A_and_F) {
+  double present_value = agent_->PV(10, 0.05, 10000.0, 100.0);
+  double pv_hand_calc = 6911.306028;
+  EXPECT_NEAR(present_value, pv_hand_calc, 1e-3);
+}
+
+TEST_P(AgentTests, PV_only_F) {
```
Comment (gonuke) [inline tests/agent_tests/agent_tests.cc:88]: You can probably combine a few different tests into a single function:
* PV_only_F
* PV_only_A
* PV_A_and_F

Could also try a few different values of i and n.  So `test_PV` could be:
* i=0, n=0, F=1, A=0
* i>0, n=1, F=1, A=0
* i>0, n>1, F=1, A=0
* i>0, n>1, F>1, A=0
* i>0, n>1, F=0, A=1
* i>0, n>1, F=0, A>1
* i>0, n>1, F>1, A>1
Etc

It's also valid to test them against a computed response rather than a hand calculated response. E.g.
`pv(2,0.01,1,0) == pow(1.01,2)`

**Review by gonuke — CHANGES_REQUESTED** [review]
I think it probably makes sense to discuss the desired behavior and broad testing strategy in person

**Comment by dean-krueger** [comment]
Sounds good. I'll stick a pin in this until Wednesday and make it an item for the Fuel Cycles meeting so we can talk about it.

File: tests/agent_tests/agent_tests.cc [hunk 7]
```diff
@@ -78,3 +79,109 @@ TEST_P(AgentTests, SetThenGetEconParam) {
...
+  double finance_eps = 1e-4;
+
+  EXPECT_NEAR(F, agent_->PV(n,0,F,0), finance_eps);  // discount rate = 0
+  EXPECT_NEAR(F, agent_->PV(0,i,F,0), finance_eps);  // discount time = 0
+  EXPECT_NEAR(F / (1+i), agent_->PV(1,i,F,0), finance_eps); // discount time = 1
+  EXPECT_NEAR(F / std::pow((1+i), n), agent_->PV(n,i,F,0), finance_eps); // general
+  EXPECT_NEAR(F / std::pow((1 + (i/12)), n*12), agent_->PV(n*12,i/12,F,0), finance_eps); // general
+  EXPECT_NEAR(F / std::pow((1+i), -n), agent_->PV(-n,i,F,0), finance_eps); // negative time
+
+  EXPECT_NEAR(A, agent_->PV(n,0,0,A), finance_eps); // discount rate = 0
```
Comment (gonuke) [inline tests/agent_tests/agent_tests.cc:97]: My bad:
```suggestion
  EXPECT_NEAR(A * n, agent_->PV(n,0,0,A), finance_eps); // discount rate = 0
```

File: tests/agent_tests/agent_tests.cc [hunk 8]
```diff
@@ -78,3 +79,109 @@ TEST_P(AgentTests, SetThenGetEconParam) {
...
+  EXPECT_NEAR(0, agent_->PV(0,i,0,A), finance_eps);  // discount time = 0
+  EXPECT_NEAR(A / (1+i), agent_->PV(1,i,0,A), finance_eps); // discount time = 1
+  EXPECT_NEAR(A * (1 - std::pow((1 + i), -n)) / i, agent_->PV(n,i,0,A), finance_eps); // general
+  EXPECT_NEAR(A * (1 - std::pow((1 + (i/12)), -n*12)) / (i/12), agent_->PV(n*12,i/12,0,A), finance_eps); // general
+  EXPECT_NEAR(A * (1 - std::pow((1 + i), n)) / i, agent_->PV(-n,i,0,A), finance_eps); // negative time
+
+  EXPECT_NEAR(F + A, agent_->PV(n,0,F,A), finance_eps); // discount rate = 0
```
Comment (gonuke) [inline tests/agent_tests/agent_tests.cc:104]: Same error as above
```suggestion
  EXPECT_NEAR(F + A * n, agent_->PV(n,0,F,A), finance_eps); // discount rate = 0
```

File: tests/agent_tests/agent_tests.cc [hunk 9]
```diff
@@ -78,3 +79,109 @@ TEST_P(AgentTests, SetThenGetEconParam) {
...
+  int n = 5;
+  double i = 0.1;
+  double P = 1;
+  double A = 0.5;
+  double finance_eps = 1e-4;
+
+  // testing by "round trip" with already tested functions
+  EXPECT_NEAR(P, agent_->PV(n,0,agent_->FV(n,0,P,0),0), finance_eps);  // discount rate = 0
+  EXPECT_NEAR(P, agent_->PV(0,i,agent_->FV(0,i,P,0),0), finance_eps);  // discount time = 0
+  EXPECT_NEAR(P, agent_->PV(i,i,agent_->FV(1,i,P,0),0), finance_eps);  // discount time = 1
```
Comment (gonuke) [inline tests/agent_tests/agent_tests.cc:124]: type
```suggestion
  EXPECT_NEAR(P, agent_->PV(1,i,agent_->FV(1,i,P,0),0), finance_eps);  // discount time = 1
```

File: tests/agent_tests/agent_tests.cc [hunk 10]
```diff
@@ -78,3 +79,109 @@ TEST_P(AgentTests, SetThenGetEconParam) {
...
+  EXPECT_NEAR(P, agent_->PV(n,i,agent_->FV(n,i,P,0),0), finance_eps);  // general
+  EXPECT_NEAR(P, agent_->PV(n*12,i/12,agent_->FV(n*12,i/12,P,0),0), finance_eps);  // general
+  EXPECT_NEAR(P, agent_->PV(-n,i,agent_->FV(-n,i,P,0),0), finance_eps);  // general
+  
+  // manual testing
+  EXPECT_NEAR(A, agent_->FV(n,0,0,A), finance_eps); // discount rate = 0
```
Comment (gonuke) [inline tests/agent_tests/agent_tests.cc:130]: repeated error
```suggestion
  EXPECT_NEAR(A * n, agent_->FV(n,0,0,A), finance_eps); // discount rate = 0
```

File: tests/agent_tests/agent_tests.cc [hunk 11]
```diff
@@ -78,3 +79,109 @@ TEST_P(AgentTests, SetThenGetEconParam) {
...
+  EXPECT_NEAR(0, agent_->FV(0,i,0,A), finance_eps);  // discount time = 0
+  EXPECT_NEAR(A, agent_->FV(1,i,0,A), finance_eps); // discount time = 1
+  EXPECT_NEAR(A * (std::pow((1 + i), n) - 1) / i, agent_->PV(n,i,0,A), finance_eps); // general
+  EXPECT_NEAR(A * (std::pow((1 + i/12), n*12) - 1) / i/12, agent_->PV(n*12,i/12,0,A), finance_eps); // general
+  EXPECT_NEAR(A * (std::pow((1 + i), -n) - 1) / i, agent_->PV(-n,i,0,A), finance_eps); // general
```
Comment (gonuke) [inline tests/agent_tests/agent_tests.cc:135]: Dumb copy/paste errors on my part
```suggestion
  EXPECT_NEAR(A * (std::pow((1 + i), n) - 1) / i, agent_->FV(n,i,0,A), finance_eps); // general
  EXPECT_NEAR(A * (std::pow((1 + i/12), n*12) - 1) / i/12, agent_->FV(n*12,i/12,0,A), finance_eps); // general
  EXPECT_NEAR(A * (std::pow((1 + i), -n) - 1) / i, agent_->FV(-n,i,0,A), finance_eps); // general
```

File: tests/agent_tests/agent_tests.cc [hunk 12]
```diff
@@ -78,3 +79,109 @@ TEST_P(AgentTests, SetThenGetEconParam) {
...
+
+  EXPECT_NEAR(P + A, agent_->FV(n,0,P,A), finance_eps); // discount rate = 0
```
Comment (gonuke) [inline tests/agent_tests/agent_tests.cc:137]: repeated error
```suggestion
  EXPECT_NEAR(P + A * n, agent_->FV(n,0,P,A), finance_eps); // discount rate = 0
```

File: tests/agent_tests/agent_tests.cc [hunk 13]
```diff
@@ -78,3 +79,109 @@ TEST_P(AgentTests, SetThenGetEconParam) {
...
+  EXPECT_NEAR(P, agent_->FV(0,i,P,A), finance_eps); // discount time = 0
+  EXPECT_NEAR(P * (1+i) + A, agent_->FV(1,i,P,A), finance_eps); // discount time = 1
+  EXPECT_NEAR(P * std::pow((1+i), n) + A * (std::pow((1 + i), n) - 1),
```
Comment (gonuke) [inline tests/agent_tests/agent_tests.cc:140]:
```suggestion
  EXPECT_NEAR(P * std::pow((1+i), n) + A * (std::pow((1 + i), n) - 1) / i,
```

File: tests/agent_tests/agent_tests.cc [hunk 14]
```diff
@@ -78,3 +79,109 @@ TEST_P(AgentTests, SetThenGetEconParam) {
...
+  int n = 5;
+  double i = 0.1;
+  double P = 1.5;
+  double F = 3.5;
+  double finance_eps = 1e-4;
+
+  // all testing by "round trip" with already tested functions
+
+  EXPECT_NEAR(P, agent_->PV(n,0,0,agent_->PMT(n,0,P,0)), finance_eps); // discount rate = 0
+  EXPECT_NEAR(P, agent_->PV(0,i,0,agent_->PMT(0,i,P,0)), finance_eps); // discount time = 0
```
Comment (gonuke) [inline tests/agent_tests/agent_tests.cc:156]: The left hand side evaluates to `nan` - this case may not be well defined and we should discuss the right solution.

Basically, there is no correct answer for how big the equivalent payment is when there are 0 payment.  So maybe this is a case where we want to make sure n >= 0 and otherwise throw an exception.

File: tests/agent_tests/agent_tests.cc [hunk 15]
```diff
@@ -78,3 +79,109 @@ TEST_P(AgentTests, SetThenGetEconParam) {
...
+  EXPECT_NEAR(P, agent_->PV(1,i,0,agent_->PMT(1,i,P,0)), finance_eps); // discount time = 1
+  EXPECT_NEAR(P, agent_->PV(n,i,0,agent_->PMT(n,i,P,0)), finance_eps); // general
+  
+  EXPECT_NEAR(F, agent_->FV(n,0,0,agent_->PMT(n,0,0,F)), finance_eps); // discount rate = 0
+  EXPECT_NEAR(F, agent_->FV(0,i,0,agent_->PMT(0,i,0,F)), finance_eps); // discount time = 0
```
Comment (gonuke) [inline tests/agent_tests/agent_tests.cc:161]: same error as above with `nan`

File: tests/agent_tests/agent_tests.cc [hunk 16]
```diff
@@ -78,3 +79,109 @@ TEST_P(AgentTests, SetThenGetEconParam) {
...
+  std::vector<double> A2({1.0,2.0,3.0,3.5,4.71});
+
+  int n = A1.size();
+  double i = 0.1;
+  double finance_eps = 1e-4;
+
+  EXPECT_NEAR(agent_->PV(n,i,0,A1[0]), agent_->PV(i, A1), finance_eps);
+  double PV_A2 = 0;
+  for (int n2=0; n2 < A2.size(); n2++) {
+    PV_A2 += agent_->PV(n2+1, i, 0, A2[n2]);
```
Comment (gonuke) [inline tests/agent_tests/agent_tests.cc:182]: Each payment acts like a separate single future value in this case
```suggestion
    PV_A2 += agent_->PV(n2+1, i, A2[n2], 0);
```

**Review by gonuke — CHANGES_REQUESTED** [review]
Now that I've actually seen the tests run (my bad for not doing that before), I've found all of my own errors here.

**Comment by dean-krueger** [comment]
Well that's a convenient development! I was just about to go dig through these myself, but thanks for taking another look at them. I'll make these changes and then see where we stand with the tests.

File: src/economic_entity.h [hunk 17]
```diff
@@ -32,6 +32,106 @@ class EconomicEntity {
...
+    }
+
+    double p_term = P;
+    double f_term = F / std::pow((1 + i), n);
+    
+    // Since we're allowing certain terms to be zero
+    if (i != 0) {
+        return (p_term + f_term) * i / (1 - std::pow((1 + i), -n));
+    } else {
+        return (p_term + F) / n;
```
[the remark this answers is not in the document]
Comment (gonuke) [reply src/economic_entity.h:106]: Unless you reject this suggestion, I'll merge when this has been committed

**Review by gonuke — APPROVED** [review]
Thanks @dean-krueger
